tasmota/tasmota_xsns_sensor/xsns_02_analog.ino

```diff
@@ -417,6 +431,76 @@ float AdcGetMq(uint32_t idx) {
   return ppm;
 }
 
+void AddSampleTds(uint32_t idx){
+  AddLog(LOG_LEVEL_DEBUG, PSTR(D_LOG_APPLICATION "Adding sample for tds-sensor"));
+  
+  int _adc = AdcRead(Adc[idx].pin, 2);
+
+  // init af array at same value
+  if (Adc[idx].indexOfPointer==-1)
+  {
+    for (int i = 0; i < ANALOG_TDS_SAMPLES; i ++) 
+      Adc[idx].tds_samples[i] = _adc;
+  }
+  else
+    Adc[idx].tds_samples[Adc[idx].indexOfPointer] = _adc;
+  Adc[idx].indexOfPointer++;
+  if (Adc[idx].indexOfPointer==ANALOG_TDS_SAMPLES)
+    Adc[idx].indexOfPointer=0;
+  
+}
+
+// median filtering algorithm
+int getMedianNum(int bArray[], int iFilterLen){
+  int bTab[iFilterLen];
+  for (byte i = 0; i<iFilterLen; i++)
+  bTab[i] = bArray[i];
+  int i, j, bTemp;
+  for (j = 0; j < iFilterLen - 1; j++) {
+    for (i = 0; i < iFilterLen - j - 1; i++) {
+      if (bTab[i] > bTab[i + 1]) {
+        bTemp = bTab[i];
+        bTab[i] = bTab[i + 1];
+        bTab[i + 1] = bTemp;
+      }
+    }
+  }
+  if ((iFilterLen & 1) > 0){
+    bTemp = bTab[(iFilterLen - 1) / 2];
+  }
+  else {
+    bTemp = (bTab[iFilterLen / 2] + bTab[iFilterLen / 2 - 1]) / 2;
+  }
+  return bTemp;
+}
+
+float AdcGetTds(uint32_t idx) {
+  AddLog(LOG_LEVEL_DEBUG, PSTR(D_LOG_APPLICATION "Getting value for tds-sensor "));
+  
+  
+  float temp = (float)Adc[idx].param1;
+
+  temp = ANALOG_TDS_COMPENSATION_TEMPERATURE; // TODO:  Adc[idx].param1 is not init with ANALOG_TDS_COMPENSATION_TEMPERATURE
+  // it would be better if we can read temperature sensor directly for compensation
+
+  int median = getMedianNum(Adc[idx].tds_samples, ANALOG_TDS_SAMPLES);
+  float voltage = median * ANALOG_V33 / ((FastPrecisePow(2, ANALOG_RESOLUTION)) - 1);
```

Comment:
It is inefficient to use Pow function for a power of 2. Especially for a constant result
Please use (1 << ANALOG_RESOLUTION)

Reply:
Also beware that `FastPrecisePow` is horribly imprecise outside of intended use cases. 
> 2 to the power of 2 is 3.884

Reply:
Thanks for info. I use shift left operation in this case.

I need little improvment of this sensor. Actualy TDS need temperature compensaton for precise value.  It would be the best that we read temperature from sensor connected on device.  

What is best practice to read read temperature sensor data?

Reply:
Use global temperature variable set by attached temp sensor or by mqtt message.

Reply:
Thanks Theo for very good hint. 

The version which read global temperature is commited.
But with global temperature variable are some issues.
In my case global temperature variable switching between external temperature sensor and CPU temperature sensor.

Here is log export from my device:
00:02:16.846 APP: Ppm read. ADC-RAW: 0.00, ppm: 0.00, temp: 29.12
00:02:17.081 APP: Getting value for tds-sensor 
00:02:17.086 APP: Ppm read. ADC-RAW: 0.00, ppm: 0.00, temp: 53.33


First entry with value 29.12 is external temperature sensor. The last value (just next tick) 53.33 is the CPU teprature. 


I found that is this declared in file xsns_127_esp32_sensors.ino (line 64). There are variable add_global_temp 
which check in json string exist any temperature parameter. This solution is unreliable.
Maybe is good enough that we never define CPU temperature as global temperature variable (We more offen need external temperature in sensor API). 
Or better and more comprehensive solution is make UI where user select Global temperature sensor name.

Can you fix this issiue with global variable?

Reply:
If température is needed, I would imagine that the water temperature is needed, not the ambiant air.

Reply:
Was fixed recently.

Latest global temp/hum/press change makes it possible for a user to select a temperature from the JSON message. In your case you insert a manual temp. Global temperatures timeout after 5 minutes and become unavailable unless a new update is provided (either by a local temp sensor or externally by MQTT).

So to fix your issue you'll need to insert a real temperature at least once every 5 minutes. If you do not do this the next available temperature sensor is used. In your case the internal ESP32 one. Just disable the ESP32 sensor with command ``setsensor127 0``.

Best approach is to have a local temp sensor which will provide temp in the JSON message which in turn can be used as global temp with command ``GlobalTemp2 1`` signalling the first JSON temperature in the message.